04-spatial-operations.Rmd

```diff
@@ -202,13 +205,15 @@ plot(joined["Population (millions)"], add = TRUE)
 par(mfrow = c(1, 1))
 ```
 
-Like attribute data aggregation, covered in section \@ref(vector-attribute-aggregation), spatial data aggregation can be a way of *condensing* data.
+<!-- I would expand it a little bit on spatial joins. Typically, this operation is also known as spatial overlay. Of course, you can also use other geometries (lines), and especially you can not only use st_intersects (another reason why the topological stuff should have been already covered before). Also interesting would be to show what happens if there are multiple matches, e.g., multiple points per polygon. In sp, this was solved using a list containing the IDs of the matched objects while not returning the corresponding attribute table. Overall, a comparison with sp::over would be also interesting -->
+
+Like attribute data aggregation, covered in section \@ref(vector-attribute-aggregation), spatial data aggregation (also known as dissolving polygons) can be a way of *condensing* data.
 Aggregated data show some statistic about a variable (typically mean average or total) in relation to some kind of *grouping variable*. 
 For attribute data aggregation the grouping variable is another variable, typically one with few unique values relative to the number of rows.
 The `continent` variable in the `world` dataset is a good example:
-there are only 8 unique continents but 177 countries.
-In section \@ref(vector-attribute-aggregation) the aggregation process condensed the `world` dataset down into only 8 rows and an aggregated `pop` variable representing the total population per continent (see Figure \@ref(fig:continent-pop)).
-
+there are only eight unique continents but 177 countries.
+In section \@ref(vector-attribute-aggregation) the aggregation process condensed the `world` dataset down into only eight rows and an aggregated `pop` variable representing the total population per continent (see Figure \@ref(fig:continent-pop)).
+<!-- CAREFUL, in fact, fig:continent-pop has already been a spatial operation (dissolving polygons). You should have shown the table only in the previous section --> 
```

Comment:
Be careful! I guess fig:continent-pop belongs here because it was a spatial operation! We should have just shown the table in c3!

Reply:
OK - could you do that?

Reply:
Sure, there's only one problem. Maybe you remember, we have discussed this before: Dissolving the world led to some strange artifacts in the output (borders within continents where no borders should be). I have checked this is due to an unclean topology, here overlapping polygons. I have tried to clean this using GRASS. This makes it better but not perfect. So the question is if we cannot use a clean world map in the first place? But maybe that's not that easy, I have tested also other world maps available in R and these were also unclean... We can also use just another dataset, e.g.,

```r
library(spData)
group_by(us_states, REGION) %>% 
   summarize(pop = sum(total_pop_15)) %>% 
   select(pop) %>%
   plot
```

Reply:
I see your point:

``` r
library(spData)
library(sf)
#> Linking to GEOS 3.5.1, GDAL 2.2.1, proj.4 4.9.2, lwgeom 2.3.3 r15473
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
group_by(world, continent) %>% summarize(pop = sum(pop)) %>% select(pop) %>% 
  plot
```

![](https://i.imgur.com/ZPXJJA0.png)

Reply:
Using the US data instead sounds fine but we should mention how to fix it.

Reply:
But we should also so how to fix it, as illustrated below (thanks @Nowosad - pls link to issue):

``` r
library(spData)
library(sf)
#> Linking to GEOS 3.5.1, GDAL 2.2.1, proj.4 4.9.2, lwgeom 2.3.3 r15473
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
world %>% st_set_precision(precision = -1) %>% group_by(continent) %>% summarize(pop = sum(pop)) %>% 
  select(pop) %>% plot
```

![](https://i.imgur.com/nS9GC2V.png)

Reply:
Well, that's only a quick-and-dirty solution since this doesn't clean the topology, and overlapping polygons still remain 😄 . But yes, good point, we should mention, that setting the precision sometimes facilitates geometrical operations with GEOS.

Reply:
@jannes-m The question is if there is a good way to clean a topology in R? Maybe it could be a nice small section in the book.

Reply:
Depends on the input, as usual, see also Barry's answer to this [post](http://r-sig-geo.2731867.n2.nabble.com/Cleaning-up-self-intersections-td7581945.html). 
My usual approach is to use rgrass7. `writeVECT`  uses `v.in.ogr` and this cleans the vector geometry and builds a topology. And then I simply load the cleaned data back into R using `readVECT`. In many cases this works, however, if the input data is too unclean, then GRASS can't solve this either easily. This was the case with the `world` dataset.